torch/_dynamo/_trace_wrapped_higher_order_op.py

```diff
@@ -0,0 +1,109 @@
+import functools
+
+import torch.utils._pytree as pytree
+from torch._C import DispatchKey
+from torch._higher_order_ops.utils import autograd_not_implemented
+
+from torch._ops import HigherOrderOperator
+from torch._subclasses import FakeTensorMode
+
+from torch.fx.experimental.proxy_tensor import ProxyTorchDispatchMode, track_tensor_tree
+from torch.utils._python_dispatch import _get_current_dispatch_mode
+
+
+# Trace wrapped is a higher order op meant for both invoking a bound function,
+# and for registering it as a call_function in the backward graph.
+# This allows us to re-enter dynamo during compiled autograd to trace (or graph break)
+# the functions as needed. This, in turn, means we can support functions in backward with complex python
+# state mutation. If we were to not do this, the functions would get inlined into their composing aten ops,
+# and we would lose the python state mutation.
+def _trace_wrapped(*args, fn):
+    return _trace_wrapped_op(*args, fn=fn)
+
+
+_trace_wrapped_op = HigherOrderOperator("_trace_wrapped")
+
+
+def self_invoke(*args, fn):
+    # This wrapper intercepts calls to fn, and calls the real fn via _trace_wrapped
+    # Dynamo unpacks this higher order op into calling the wrapped fn
+    return _trace_wrapped_op(*args, fn=fn)
+
+
+@_trace_wrapped_op.py_impl(ProxyTorchDispatchMode)
+def inner_trace(mode, *args, fn):
+    import torch._functorch.aot_autograd
+
+    assert len(args) == 1, breakpoint()
+    grad = args[0]
+    assert isinstance(grad, torch.Tensor)
+
+    if isinstance(fn, functools.partial):
+        fn.__name__ = fn.func.__name__  # type: ignore[attr-defined]
+    original_fn = fn
+    # We've implemented a higher-order operator that remains consistent in proxy tensor tracing.
+    # However, Dynamo is aware and traces this into its genuine functionality.
+    # The operator's purpose is to facilitate invoking non-traceable functions
+    # and embedding them directly into the graph. Essentially, this transforms function
+    # calls into "leaf modules" as per traditional FX terminology.
+    # Note: Instead of naming it "allow_in_graph", we opted for a different name since "allow_in_graph"
+    # might imply that it's traceable, whereas this function is intrinsically non-traceable.
+    # Note2: I hate this name
+    fn = functools.partial(self_invoke, fn=fn)
+    fn.__name__ = fn.func.__name__
+
+    proxy_args = pytree.tree_map(mode.tracer.unwrap_proxy, (grad,))
+    out_proxy = mode.tracer.create_proxy(
+        "call_function", fn, proxy_args, {}, name="invocation"
+    )
+    grad = track_tensor_tree(grad, out_proxy, constant=None, tracer=mode.tracer)
+
+    # We have a little shortcut here, wherein we DO NOT yet run a meta func, and so
```

Comment:
No action needed but just thinking: do you think we should be worried that this will prevent fusions? E.g.:
```
grad_inp = last_op_in_backward_hook(grad_out)
grad_inp_updated = functional_assert_tensor_metadata(grad_inp, ...)
gradient = next_bw_node(grad_inp_updated)
```

Before inductor could fuse between the end of the hook and the beginning of the next node, but now we need to materialize the output of the hook so we can assert on it.

Is the longer term plan to arrange for a meta function to run instead (where I guess we need a combo of dynamo + functionalization to make sure that plumbing fake tensors through the hook doesn't cause side effects?)

Reply:
Hmm, I am not sure, this is a good question.

Reply:
Assuming that dynamo + functionalization have already run on the hook by the time we’re adding it to the graph, what’s preventing us from plumbing fake tensors through it?

Reply:
Well, at this point, you still are holding on to the raw user fn. You dont want to invoke it for meta prop, even with fake tensors, because you might mutate python state.

Reply:
yeah :( nvm.

Still just thinking: maybe we should put a "TODO backward guards" here? (and yet another reason to implement backward guards at some point). The situation here seems kinda similar: let's say we optimistically guessed the output metadata == input metadata. Will we have done anything incorrect with that information by the time we hit the backward?

(1) traced forward graph: this will still be correct, the the hook is in the backward
(2) partitioning: getting the sizes of the outputs to the hook wrong won't change any partitioning decisions: the partitioner will have put all the outputs of the hook in the backward graph anyway
(3) Traced backward graph: this will be wrong, and we'd have to retrace.

In theory we could probably have dynamo fail some guards when it tries to trace through the hooks during the backward, and realize that it needs to retrace the entire backward (this time, actually plumbing fake tensors through the dynamo'd hook so we get correct shapes when tracing out the backward).

Reply:
Well... we insert a runtime assert - so if nothing else, this fails. We do have guards for compiled autograd though, right?

Reply:
check https://github.com/pytorch/pytorch/pull/109690/files#r1333134540

Reply:
Discussed offline, updated to match